gsl_projet/tests/services/test_projet_services.py

```diff
@@ -286,3 +291,84 @@ def test_get_dotations_from_field(field, value, expected_dotation):
+
+
+@pytest.fixture
+def projet():
+    return ProjetFactory()
+
+
+@pytest.mark.django_db
+def test_update_dotation_with_no_value(projet, caplog):
+    with caplog.at_level(logging.WARNING):
+        ProjetService.update_dotation(projet, [])
+    assert f"Projet {projet.__str__()} must have at least one dotation" in caplog.text
+    assert projet.dotations == []
+
+
+@pytest.mark.django_db
+def test_update_dotation_with_more_than_2_values(projet, caplog):
+    with caplog.at_level(logging.WARNING):
+        ProjetService.update_dotation(projet, [DOTATION_DETR, DOTATION_DSIL, "unknown"])
+    assert (
+        f"Projet {projet.__str__()} can't have more than two dotations" in caplog.text
+    )
+    assert projet.dotations == []
+
+
+@pytest.mark.parametrize("dotation", [DOTATION_DETR, DOTATION_DSIL])
+@mock.patch.object(
+    DotationProjetService, "create_simulation_projets_from_dotation_projet"
+)
+@pytest.mark.django_db
+def test_update_dotation_from_one_dotation_to_another(
+    mock_create_simulation_projets, dotation, projet
+):
+    original_dotation_projet = DotationProjetFactory(projet=projet, dotation=dotation)
+    SimulationProjetFactory.create_batch(3, dotation_projet=original_dotation_projet)
+    ProgrammationProjetFactory.create(dotation_projet=original_dotation_projet)
+
+    new_dotation = DOTATION_DSIL if dotation == DOTATION_DETR else DOTATION_DETR
+    ProjetService.update_dotation(projet, [new_dotation])
+
+    assert projet.dotations == [new_dotation]
+    assert projet.dotationprojet_set.count() == 1
+    dotation_projet = projet.dotationprojet_set.first()
+
+    assert mock_create_simulation_projets.call_count == 1
+    mock_create_simulation_projets.assert_called_once_with(dotation_projet)
+
+    # Check that the old dotation_projet is deleted
+    assert DotationProjet.objects.filter(pk=original_dotation_projet.pk).count() == 0
+    assert SimulationProjet.objects.count() == 0
+    assert ProgrammationProjet.objects.count() == 0
+
+
+@pytest.mark.parametrize("original_dotation", [DOTATION_DETR, DOTATION_DSIL])
+@mock.patch.object(
+    DotationProjetService, "create_simulation_projets_from_dotation_projet"
+)
+@pytest.mark.django_db
+def test_update_dotation_from_one_to_two(
+    mock_create_simulation_projets, original_dotation, projet
+):
+    original_dotation_projet = DotationProjetFactory(
+        projet=projet, dotation=original_dotation
+    )
+    SimulationProjetFactory.create_batch(3, dotation_projet=original_dotation_projet)
+    ProgrammationProjetFactory.create(dotation_projet=original_dotation_projet)
+
+    ProjetService.update_dotation(projet, [DOTATION_DETR, DOTATION_DSIL])
+
+    assert projet.dotationprojet_set.count() == 2
+    for dotation in projet.dotations:
+        assert dotation in [DOTATION_DETR, DOTATION_DSIL]
+    dotation_projets = projet.dotationprojet_set.all()
+
+    new_dotation_projet = dotation_projets.exclude(
+        pk=original_dotation_projet.pk
+    ).first()
```

Comment:
On n'a pas besoin de la variable intermédiaire dont le nom me semble poins riche en info que `projet.dotationprojet_set` : 

```suggestion
    new_dotation_projet = projet.dotationprojet_set.exclude(
        pk=original_dotation_projet.pk
    ).first()
```